### DISCUSSION TITLE
(My)SQL schema discussion

### DISCUSSION THREAD

**Discussion by jessepeterson** [micromdm/nanomdm#3] [General]
I'd like to start some discussion about the decisions made with the MySQL (and potentially other RDBMS) schemas. Specifically about column types and table relationships.

I pointed out some things in the `schema.sql` under the storage backend in NanoMDM. But things like: to binary-column or not (leaning not). Which primary keys to use for command responses (currently just enrollment ID and result status). Check out the schema file for more info.

Otherwise just a general discussion about SQL schema as related to MDM and NanoMDM.

**Comment by jessepeterson** [comment]
The device queue in NanoMDM's mysql storage backend uses an SQL VIEW to query the next command for a device. To consider folding this VIEW back into a real query instead of a view.

Reasons for a VIEW:

* Can replace the view with your own specific VIEW if you so pleased without changing code
* Matches the same queries that devices do when using the queue manually (e.g. for debugging or inspection)
* Consistent query for future API usage (e.g. command/queue inspection) — see #6

Reasons against a VIEW:

* Just more complexity/another abstraction to manage (including performance/indexing etc.)
* Also hides complexity in the code of the view (could also be a positive)

And I'm sure more reasons both for and against. 

VIEW use in code:

https://github.com/micromdm/nanomdm/blob/fce958e750f512532e9a6aeb9f57227d192087f3/storage/mysql/queue.go#L89

View in schema:

https://github.com/micromdm/nanomdm/blob/fce958e750f512532e9a6aeb9f57227d192087f3/storage/mysql/schema.sql#L217-L239

The idea here is to eliminate the VIEW usage in the queue code (essentially just embed the VIEW query into `queue.go`. The VIEW itself in the schema can stay it just won't be used by nanomdm code.

Thoughts?

**Reply by tomaswallentinus** [reply]
Is there a reason not to use a stored procedure instead? That way you have cache as well.

**Reply by jessepeterson** [reply]
> Is there a reason not to use a stored procedure instead? That way you have cache as well.

I think the issue we want to go the _other_ way (removing database-specific complexities from core code). In other words: think about environments that can't or won't support stored procedures or VIEWs, etc. Just plain ol' CRUD queries. (Even think about environments that don't allow foreign key constraints).

**Reply by jessepeterson** [reply]
Removal of the VIEW from the MySQL backend was implemented in #59 FYI.

**Comment by jessepeterson** [comment]
See discussion about queue management (i.e. command deletion) over in #43.
